erigon-lib/kv/mdbx/kv_mdbx.go

```diff
@@ -2094,58 +2142,88 @@ type cursorDup2iter struct {
 	ctx                         context.Context
 }
 
-func (s *cursorDup2iter) init(table string, tx kv.Tx) (*cursorDup2iter, error) {
+func (s *cursorDup2iter) init(table string, tx kv.Tx) error {
 	if s.orderAscend && s.fromPrefix != nil && s.toPrefix != nil && bytes.Compare(s.fromPrefix, s.toPrefix) >= 0 {
-		return s, fmt.Errorf("tx.Dual: %x must be lexicographicaly before %x", s.fromPrefix, s.toPrefix)
+		return fmt.Errorf("tx.Dual: %x must be lexicographicaly before %x", s.fromPrefix, s.toPrefix)
 	}
 	if !s.orderAscend && s.fromPrefix != nil && s.toPrefix != nil && bytes.Compare(s.fromPrefix, s.toPrefix) <= 0 {
-		return s, fmt.Errorf("tx.Dual: %x must be lexicographicaly before %x", s.toPrefix, s.fromPrefix)
+		return fmt.Errorf("tx.Dual: %x must be lexicographicaly before %x", s.toPrefix, s.fromPrefix)
 	}
 	c, err := tx.CursorDupSort(table)
 	if err != nil {
-		return s, err
+		return err
 	}
 	s.c = c
 	k, _, err := c.SeekExact(s.key)
 	if err != nil {
-		return s, err
+		return err
 	}
 	if k == nil {
-		return s, nil
+		return nil
 	}
 
 	if s.fromPrefix == nil { // no initial position
 		if s.orderAscend {
 			s.nextV, err = s.c.FirstDup()
+			if err != nil {
+				return err
+			}
 		} else {
 			s.nextV, err = s.c.LastDup()
+			if err != nil {
+				return err
+			}
 		}
-		return s, err
+		return nil
 	}
 
 	if s.orderAscend {
 		s.nextV, err = s.c.SeekBothRange(s.key, s.fromPrefix)
-		return s, err
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// to find LAST key with given prefix:
+	nextSubtree, ok := kv.NextSubtree(s.fromPrefix)
+	if ok {
+		_, s.nextV, err = s.c.SeekBothExact(s.key, nextSubtree)
+		if err != nil {
+			return err
+		}
+		_, s.nextV, err = s.c.PrevDup()
+		if err != nil {
+			return err
+		}
 	} else {
-		// seek exactly to given key or previous one
-		_, s.nextV, err = s.c.SeekBothExact(s.key, s.fromPrefix)
-		if s.nextV == nil { // no such key
-			_, s.nextV, err = s.c.PrevDup()
+		s.nextV, err = s.c.LastDup()
+		if err != nil {
+			return err
 		}
-		return s, err
 	}
+	return nil
 }
 
 func (s *cursorDup2iter) advance() (err error) {
 	if s.orderAscend {
 		_, s.nextV, err = s.c.NextDup()
+		if err != nil {
+			return err
+		}
```

Comment:
Let's keep re-formatting outside of CodeReview process - because: 
1. it's not important 
2. it's automatable by something like `golangci-lint run --no-config --fix --disable-all --enable gofumpt ./core` 
3. it's controversial - people are easy to fall in "but I like having dedicated code patch for errors - easy to add `fmt.Printf("debug")` only for error path, easy to extend by `fmt.Errorf("more context: %w", err)`" - and such discussions are irrelevant for current PR. 
4. Some people are in another timezones and for them - wakeup, see rejected PR because of re-format, fix 1 line, wait one more day for CR - is frustrating.